Allow bounded channel to be created with drop delegate

Temppus opened the pull request:
Fixes #36522 
Add additional CreateBounded overload with delegate parameter that will be called when item is being dropped from channel
Added unit tests

Temppus committed:
Bounded channel can be created with drop delegate.

- Add additional CreateBounded overload with delegate parameter that will be called when item is being dropped from channel
- Added unit tests

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=50331) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=50331) before we can accept your contribution.<br/><br/>:x: Temppus [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=50331)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=50331) it.</sub>

Temppus committed:
Fix typo in comment.

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -393,6 +397,7 @@ public override bool TryWrite(T item)
                     {
                         // The channel is full.  Just ignore the item being added
                         // but say we added it.
+                        _parent._itemDropped?.Invoke(item);
```
We shouldn't invoke the delegate while the lock is held.  Same for all the case below.

Temppus replied:
This is kinda tricky because I have to:
- store item that should be dropped outside of lock statement
- item is type of T which can be either struct/class, so I can not use default keyword here to signal whether some item should be dropped. I created helper struct that wraps item and has bool value signaling if struct has some item to be dropped.
- there are lot of return statements inside that lock block, which forced me to wrap it with try-finally and do invocation of delegate there

I am opened to suggestion, but currently this is best solution that came up to me.

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.cs:
```diff
@@ -49,7 +49,22 @@ public static Channel<T> CreateBounded<T>(BoundedChannelOptions options)
 
-            return new BoundedChannel<T>(options.Capacity, options.FullMode, !options.AllowSynchronousContinuations);
+            return new BoundedChannel<T>(options.Capacity, options.FullMode, !options.AllowSynchronousContinuations, itemDropped: null);
+        }
```
This method should just delegate to CreateBounded(options, null).

Temppus replied:
Done.

stephentoub commented on src/libraries/System.Threading.Channels/tests/BoundedChannelTests.cs:
```diff
@@ -17,6 +19,17 @@ protected override Channel<T> CreateFullChannel<T>()
 
+        public static IEnumerable<object[]> ChannelDropModes()
+        {
+            foreach (var mode in Enum.GetValues(typeof(BoundedChannelFullMode)).Cast<BoundedChannelFullMode>())
```
```suggestion
            foreach (BoundedChannelFullMode mode in Enum.GetValues(typeof(BoundedChannelFullMode)))
```

stephentoub commented on src/libraries/System.Threading.Channels/tests/BoundedChannelTests.cs:
```diff
@@ -249,6 +262,94 @@ public void WriteAsync_TryRead_Many_Ignore(int bufferedCapacity)
+        [MemberData(nameof(ChannelDropModes))]
+        public void DroppedDelegateIsNull_SyncWrites(BoundedChannelFullMode boundedChannelFullMode)
+        {
+            var c = Channel.CreateBounded<int>(new BoundedChannelOptions(1) { FullMode = boundedChannelFullMode }, itemDropped: null);
```
```suggestion
            Channel<int> c = Channel.CreateBounded<int>(new BoundedChannelOptions(1) { FullMode = boundedChannelFullMode }, itemDropped: null);
```

stephentoub commented on src/libraries/System.Threading.Channels/tests/BoundedChannelTests.cs:
```diff
@@ -249,6 +262,94 @@ public void WriteAsync_TryRead_Many_Ignore(int bufferedCapacity)
+        [MemberData(nameof(ChannelDropModes))]
+        public async Task DroppedDelegateIsNull_AsyncWrites(BoundedChannelFullMode boundedChannelFullMode)
+        {
+            var c = Channel.CreateBounded<int>(new BoundedChannelOptions(1) { FullMode = boundedChannelFullMode }, itemDropped: null);
```
```suggestion
            Channel<int> c = Channel.CreateBounded<int>(new BoundedChannelOptions(1) { FullMode = boundedChannelFullMode }, itemDropped: null);
```

Temppus committed:
Apply suggestions from code review

Co-authored-by: Stephen Toub <stoub@microsoft.com>

Temppus committed:
Call drop delegate outside of lock statement.

Temppus committed:
Use overload of CreateBounded method instead of calling ctor directly.

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -283,6 +287,23 @@ internal BoundedChannelWriter(BoundedChannel<T> parent)
+                        Item = itemToDrop,
+                    };
+                }
+            }
```
This struct shouldn't be necessary.  You can just have a local bool and a local T.  Adding such a struct just adds additional pressure on the JIT.

Temppus replied:
Removed

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -332,83 +353,98 @@ public override bool TryWrite(T item)
+                    {
+                        _parent._itemDropped?.Invoke(dropitem.Item);
                     }
                 }
```
I'm not in love with the idea of adding the extra try/finally, as it adds complexity and overhead.  A couple of alternatives:
1. Change the lock to be a try/finally, and just do the invocation as part of that try/finally after the call to Monitor.Exit.
2. Change the lock to be a try/finally, but also Exit the monitor prior to each return statement, then invoking the callback between the exit and the return, e.g.
```C#
bool releaseLock = false;
try
{
    Monitor.Enter(ref releaseLock);
    ...
    T droppedItem = parent._mode == BoundedChannelFullMode.DropNewest ?
        parent._items.DequeueTail() :
        parent._items.DequeueHead();
    parent._items.EnqueueTail(item);
    Monitor.Exit(parent.SyncObj);
    releaseLock = false;
    _parent._itemDropped?.Invoke(droppedItem);
    return true;
    ...
}
finally
{
    if (releaseLock)
    {
        Monitor.Exit(parent.SyncObj);
    }
}

Temppus replied:
Thanks.
Refactored according to suggestion.

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/Channel.cs:
```diff
@@ -49,7 +49,22 @@ public static Channel<T> CreateBounded<T>(BoundedChannelOptions options)
                 throw new ArgumentNullException(nameof(options));
             }
 
```
Delete the above argument validation.  It's now handled by the method we delegate to.

Temppus replied:
Done

Temppus committed:
Code review suggestions refactor.

Temppus committed:
Move Monitor.Enter before try and use local scoped parent variable everywhere.

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -332,7 +336,11 @@ public override bool TryWrite(T item)
                 BoundedChannel<T> parent = _parent;
-                lock (parent.SyncObj)
+
+                bool releaseLock = true;
+                Monitor.Enter(parent.SyncObj);
+
+                try
                 {
```
```suggestion
                bool releaseLock = false;
                try
                {
                    Monitor.Enter(parent.SyncObj, ref releaseLock);
```

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -393,24 +401,34 @@ public override bool TryWrite(T item)
+                            parent._items.DequeueTail() :
                             parent._items.DequeueHead();
-                        }
+
+                        Monitor.Exit(parent.SyncObj);
+                        releaseLock = false;
+                        parent._itemDropped?.Invoke(droppedItem);
+
                         parent._items.EnqueueTail(item);
```
Enqueueing must be done while the lock is held.
```suggestion
                        parent._items.EnqueueTail(item);
                        Monitor.Exit(parent.SyncObj);                        
                        releaseLock = false;
                        parent._itemDropped?.Invoke(droppedItem);
```

Temppus replied:
Done

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -492,7 +510,11 @@ public override ValueTask WriteAsync(T item, CancellationToken cancellationToken
                 BoundedChannel<T> parent = _parent;
-                lock (parent.SyncObj)
+
+                bool releaseLock = true;
+                Monitor.Enter(parent.SyncObj);
+
+                try
                 {
```
```suggestion
                bool releaseLock = false;
                try
                {
                    Monitor.Enter(parent.SyncObj, ref releaseLock);
```

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -569,24 +591,34 @@ public override ValueTask WriteAsync(T item, CancellationToken cancellationToken
+                            parent._items.DequeueTail() :
                             parent._items.DequeueHead();
-                        }
+
+                        Monitor.Exit(parent.SyncObj);
+                        releaseLock = false;
+                        parent._itemDropped?.Invoke(droppedItem);
+
                         parent._items.EnqueueTail(item);
```
```suggestion
                        parent._items.EnqueueTail(item);
                        Monitor.Exit(parent.SyncObj);
                        releaseLock = false;
                        parent._itemDropped?.Invoke(droppedItem);
```

Temppus replied:
Done

stephentoub commented on src/libraries/System.Threading.Channels/tests/BoundedChannelTests.cs:
```diff
@@ -249,6 +262,94 @@ public void WriteAsync_TryRead_Many_Ignore(int bufferedCapacity)
+            // Assert expected number of dropped items delegate calls
+            Assert.Equal(10, droppedItems.Count);
+        }
+
```
Can you add a test that validates the delegate isn't invoked when BoundedChannelFullMode.Wait is used?

stephentoub replied:
Also, a test that helps validate the lock isn't held while invoking the callback would be good, e.g. by having the callback do something like:
```C#
var mres = new ManualResetEventSlim();
ThreadPool.QueueUserWorkItem(delegate
{
    c.Writer.TryWrite(...);
    mres.Set();
});
mres.Wait();
```
If the lock was held while invoking the callback, code like the above should deadlock.

Temppus replied:
Added both test cases.

Temppus committed:
Drop delegate should not be called while sync lock is held.

Enqueuing of new item should be done while sync lock is being held.
Added additional tests.

Temppus committed:
Rerun gitlab CI.

Temppus committed:
Do not run deadlock test for bounded channels if platform do not support threading.

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -15,6 +15,8 @@ internal sealed class BoundedChannel<T> : Channel<T>, IDebugEnumerable<T>
     {
         /// <summary>The mode used when the channel hits its bound.</summary>
         private readonly BoundedChannelFullMode _mode;
+        /// <summary>The delegate that will be called when channel hits its bound and item is being dropped from channel.</summary>
```
```suggestion
        /// <summary>The delegate that will be invoked when the channel hits its bound and an item is dropped from the channel.</summary>
```

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -40,12 +42,14 @@ internal sealed class BoundedChannel<T> : Channel<T>, IDebugEnumerable<T>
         /// <param name="mode">The mode used when writing to a full channel.</param>
         /// <param name="runContinuationsAsynchronously">Whether to force continuations to be executed asynchronously.</param>
-        internal BoundedChannel(int bufferedCapacity, BoundedChannelFullMode mode, bool runContinuationsAsynchronously)
+        /// <param name="itemDropped">Delegate that will be called when item is being dropped from channel. See <see cref="BoundedChannelFullMode"/>.</param>
```
```suggestion
        /// <param name="itemDropped">Delegate that will be invoked when an item is dropped from the channel. See <see cref="BoundedChannelFullMode"/>.</param>
```

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -332,8 +336,13 @@ public override bool TryWrite(T item)
-                lock (parent.SyncObj)
+
+                bool releaseLock = false;
+
```
```suggestion
```

stephentoub commented on src/libraries/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs:
```diff
@@ -492,8 +512,13 @@ public override ValueTask WriteAsync(T item, CancellationToken cancellationToken
-                lock (parent.SyncObj)
+
+                bool releaseLock = false;
+
```
```suggestion
```

stephentoub commented on src/libraries/System.Threading.Channels/tests/BoundedChannelTests.cs:
```diff
@@ -249,6 +262,216 @@ public void WriteAsync_TryRead_Many_Ignore(int bufferedCapacity)
+            var secondWriteTask = c.Writer.WriteAsync(2);
+            Assert.False(secondWriteTask.IsCompleted);
+
+            // Read from chanell to free up space
```
```suggestion
            // Read from channel to free up space
```

stephentoub commented on src/libraries/System.Threading.Channels/tests/BoundedChannelTests.cs:
```diff
@@ -249,6 +262,216 @@ public void WriteAsync_TryRead_Many_Ignore(int bufferedCapacity)
+
+                dropDelegateCalled = true;
+
+                // Dropped delegate should not be called while holding the channel synchronisation lock.
```
```suggestion
                // Dropped delegate should not be called while holding the channel lock.
```

stephentoub reviewed (approved):
Thanks!

stephentoub committed:
Apply suggestions from code review
